torchvision/prototype/transforms/_geometry.py

```diff
@@ -413,76 +413,69 @@ def __init__(
 
         if self.padding is not None:
-            # update height, width with static padding data
-            padding = self.padding
-            if isinstance(padding, Sequence):
-                padding = list(padding)
-            pad_left, pad_right, pad_top, pad_bottom = F._geometry._parse_pad_padding(padding)
-            height += pad_top + pad_bottom
-            width += pad_left + pad_right
-
-        output_height, output_width = self.size
-        # We have to store maybe padded image size for pad_if_needed branch in _transform
-        input_height, input_width = height, width
+            pad_left, pad_right, pad_top, pad_bottom = self.padding
+            padded_height += pad_top + pad_bottom
+            padded_width += pad_left + pad_right
+        else:
+            pad_left = pad_right = pad_top = pad_bottom = 0
+
+        cropped_height, cropped_width = self.size
 
         if self.pad_if_needed:
-            # pad width if needed
-            if width < output_width:
-                width += 2 * (output_width - width)
-            # pad height if needed
-            if height < output_height:
-                height += 2 * (output_height - height)
-
-        if height < output_height or width < output_width:
+            if padded_height < cropped_height:
+                diff = cropped_height - padded_height
+
+                pad_top += diff
+                pad_bottom += diff
+                padded_height += 2 * diff
+
+            if padded_width < cropped_width:
+                diff = cropped_width - padded_width
+
+                pad_left += diff
+                pad_right += diff
+                padded_width += 2 * diff
+
+        if padded_height < cropped_height or padded_width < cropped_width:
             raise ValueError(
-                f"Required crop size {(output_height, output_width)} is larger then input image size {(height, width)}"
+                f"Required crop size {(cropped_height, cropped_width)} is larger than "
+                f"{'padded ' if self.padding is not None else ''}input image size {(padded_height, padded_width)}."
             )
 
-        if width == output_width and height == output_height:
-            return dict(top=0, left=0, height=height, width=width, input_width=input_width, input_height=input_height)
+        # We need a different order here than we have in self.padding since this padding will be parsed again in `F.pad`
+        padding = [pad_left, pad_top, pad_right, pad_bottom]
+        needs_pad = any(padding)
 
-        top = torch.randint(0, height - output_height + 1, size=(1,)).item()
-        left = torch.randint(0, width - output_width + 1, size=(1,)).item()
+        if padded_height > cropped_height or padded_width > cropped_width:
+            top = int(torch.randint(0, padded_height - cropped_height + 1, size=()))
+            left = int(torch.randint(0, padded_width - cropped_width + 1, size=()))
+        else:
+            top = left = 0
```

Comment:
In a previous version I checked the height and width diff individually before sampling and just setting `0` if there is no diff. This is a speed-up since in this case the call looks like `int(torch.randint(0, 1, size=()))` which is always `0`.

However, checking individually is different from what we do now in prototype and in stable:

https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/transforms/transforms.py#L634-L639

Thus, this could be regarded as BC breaking change since we would change the sampling in some cases. For example, this triggered our consistency tests since there we don't have a height diff, but call `torch.randint` twice since there is a width diff. If we check separately, we only call `torch.randint` once and thus get a different parameter set even if we set the random seed before the call.

Thoughts?

Reply:
This is a tangent comment:
https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/transforms/transforms.py#L634-L635

@vfdev-5 These are some optimizations we can do on the new Transforms to avoid calling the resize kernel if not needed. The previous implementation reaches resize which is unnecessary. Not 100% sure if this is a no-op on the C++ side but we should make sure we don't do this on the next Transforms (maybe it's fixed already?).


For the question at hand, I understand that the new implementation does fewer unnecessary random calls. Though you are right to say that this is going to make our life harder for testing the transforms (because the seeds won't be the same), I agree with Victor that there is no BC issue here. If I misunderstood your position, please let me know.

Reply:
> Though you are right to say that this is going to make our life harder for testing the transforms (because the seeds won't be the same), I agree with Victor that there is no BC issue here.

No, this is correct. I'll fix the test and revert back to separate checking the diffs.